uplink/auth.py

```diff
@@ -43,9 +50,63 @@ def __call__(self, request_builder):
         request_builder.info["headers"]["Proxy-Authorization"] = self._auth_str
 
 
-class BearerToken(object):
+class ApiTokenParam(object):
+    """
+    Authorizes requests using a token or key in a query parameter.
+    Users should subclass this class to define which parameter is the token parameter.
+    """
+    param = None
```

Comment:
I'm concerned about that approach because it mixes two different things. For an API client library built on top of uplink:

- The API library author defines the parameter to use.
- The API library consumer supplies the auth token.

I'm trying to allow the API library author to create a class (a subclass of `ApiTokenParam` or `ApiTokenHeader`) that includes only the parameter name (as a class attribute) instead of requiring the API library consumer to supply both parameter name and token.

Reply:
So, here are two forms of the subclass approach: (1) the way I envisioned it (plus a `_` to mark the param as private), and (2) using the constructor.

subclass that specifies the param as a class param
```
class ExampleApiTokenParam(ApiTokenParam):
    _param = 'api-token'
```

subclass that uses the constructor
```
class ExampleApiTokenParam(ApiTokenParam):
    def __init__(self, token):
        super(ExampleApiTokenParam, self).__init__('api-token', token)
```

Now, I'll play with a utility factory function to create such a class...

Reply:
@cognifloyd - My suggestion is attempting to accommodate both (1) the API library author (which you noted above) and (2) users who are directly using Uplink to interact with their APIs. But, does this conflicts with what you're trying to do here? As you proposed in your latter comment, we could let the API library author subclass and use the constructor to override the necessary attribute. This would let us support both the API library use case and the direct consumer. Is there another reason why you went with the first approach instead of the constructor approach?

Reply:
@cognifloyd - Disregard my above comment: I just saw that you did make the changes to expanded the generic ApiTokenParam and ApiTokenHeader classes to allow using them both directly and by subclassing. That's awesome!